### PR TITLE
Speed pattern

### PR DESCRIPTION
**Pull request by LiquidFun**
Added a new mode called SpeedPattern which shows a cube appearing, getting faster and faster, and then crashing into an imaginary wall.

For this, a new `move_all_by` method was added. Furthermore, convenience methods such as `all_on` and `all_on` were added.

### CODE REVIEW

**Review by JuliusR — APPROVED** [review]
I like it!

Feel free to merge after you checked questions and feedback in my comments.

File: src/SpeedMode.hpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        void block_slides_down() 
+        {
+            for (int z = CubePlexer::N - 1; z >= 1; --z)
+            {
+                if (cube_.is_on(0, CubePlexer::N - 1, z)) 
+                {
+                    for (int x = 0; x < CubePlexer::N; ++x)
+                    {
+                        cube_.off(x, CubePlexer::N - 1, z);
+                        cube_.on(x, CubePlexer::N - z, 0);
```
[the remark this answers is not in the document]
Comment (LiquidFun) [reply src/SpeedMode.hpp:41]: Good catch! Will fix.

File: src/main.cpp [hunk 2]
```diff
@@ -18,12 +19,15 @@ namespace T27
...
+    SpeedMode speed_mode{cube};
     PlayerMode player_mode{cube};
     FreezeMode freeze_mode{cube};
 
     auto modes = make_array<IBlinkMode *>(
         &freeze_mode // 0
         ,
+        &speed_mode // 1
+        ,
         &matrix_mode // 1
```
[the remark this answers is not in the document]
Comment (LiquidFun) [reply src/main.cpp:31]: Sounds good, forgot to change them.

File: src/SpeedMode.hpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                        if (cube_.is_on(0, 0, CubePlexer::N - 1))
+                            state = State::BlockMovesOut;
+                        else
+                            cube_.move_all_by(0, 0, 1, true);
+                        break;
+
+                    case State::BlockMovesOut:
+                        if (cube_.is_off(0, CubePlexer::N - 1, CubePlexer::N - 1)) 
+                        {
+                            state = State::BlockMovesIn;
```
[the remark this answers is not in the document]
Comment (LiquidFun) [reply src/SpeedMode.hpp:77]: It wasn't intentional at first, but I did notice it, and I think it helps sell that the cube is "off-screen" for longer. I would leave it in.

File: lib/CubePlexer/CubePlexer.hpp [hunk 4]
```diff
@@ -113,8 +117,55 @@ namespace T27
...
+        {
+            set_all_to(true);
+        }
+        void all_off()
+        {
+            set_all_to(false);
+        }
 
     private:
+        inline bool is_inside(int coord)
```
[the remark this answers is not in the document]
Comment (LiquidFun) [reply lib/CubePlexer/CubePlexer.hpp:159]: Ah, didn't know that. I'll omit it.

File: src/SpeedMode.hpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            for (int z = CubePlexer::N - 1; z >= 1; --z)
+            {
+                if (cube_.is_on(0, CubePlexer::N - 1, z)) 
+                {
+                    for (int x = 0; x < CubePlexer::N; ++x)
+                    {
+                        cube_.off(x, CubePlexer::N - 1, z);
+                        cube_.on(x, CubePlexer::N - z, 0);
+                    }
+                    return;
```
[the remark this answers is not in the document]
Comment (LiquidFun) [reply src/SpeedMode.hpp:43]: Yes, it is an unusual pattern. But I thought it solved the problem well :)

**Review by LiquidFun — COMMENTED** [review]
Will fix the problems.

**Pull request merged**